src/bvh/bvh.rs

```diff
@@ -0,0 +1,401 @@
+        let split_axis = centroid_bounds.largest_axis();
+        let split_axis_size = centroid_bounds.max[split_axis] - centroid_bounds.min[split_axis];
+
+        /// Defines a Bucket utility object.
+        #[derive(Copy, Clone)]
+        struct Bucket {
+            size: usize,
+            aabb: AABB,
+        }
+
+        impl Bucket {
+            /// Returns an empty bucket.
+            fn empty() -> Bucket {
+                Bucket {
+                    size: 0,
+                    aabb: AABB::empty(),
+                }
+            }
+
+            /// Extends this `Bucket` by the given `AABB`.
+            fn add_aabb(&mut self, aabb: &AABB) {
+                self.size += 1;
+                self.aabb = self.aabb.join(aabb);
+            }
+        }
+
+        /// Joins two `Bucket`s.
+        fn join_bucket(a: Bucket, b: &Bucket) -> Bucket {
+            Bucket {
+                size: a.size + b.size,
+                aabb: a.aabb.join(&b.aabb),
+            }
+        }
+
+        // Create six buckets, and six index assignment vectors
+        const NUM_BUCKETS: usize = 6;
+        let mut buckets = [Bucket::empty(); NUM_BUCKETS];
+        let mut bucket_assignments: [Vec<usize>; NUM_BUCKETS] = Default::default();
+
+        if split_axis_size < EPSILON {
+            // Spread the remaining shapes evenly across buckets,
+            // instead of using a heuristic
+            for (counter, idx) in (&indices).iter().enumerate() {
+                let shape = &shapes[*idx];
+                let shape_aabb = shape.aabb();
+
+                let bucket_num = counter % NUM_BUCKETS;
+
+                buckets[bucket_num].add_aabb(&shape_aabb);
+                bucket_assignments[bucket_num].push(*idx);
+            }
+        } else {
+            // Assign each shape to a bucket
+            for idx in &indices {
+                let shape = &shapes[*idx];
+                let shape_aabb = shape.aabb();
+                let shape_center = shape_aabb.center();
+
+                // Get the relative position of the shape centroid [0.0..1.0]
+                let bucket_num_relative =
+                    (shape_center[split_axis] - centroid_bounds.min[split_axis]) / split_axis_size;
+
+                // Convert that to the actual `Bucket` number
+                let bucket_num = (bucket_num_relative * (NUM_BUCKETS as f32 - 0.01)) as usize;
+
+                // Extend the selected `Bucket` and add the index to the actual bucket
+                buckets[bucket_num].add_aabb(&shape_aabb);
+                bucket_assignments[bucket_num].push(*idx);
+            }
+        }
+
+        // Compute the costs for each configuration and
+        // select the configuration with the minimal costs
+        let mut min_bucket = 0;
+        let mut min_cost = f32::INFINITY;
+        let mut child_l_aabb = AABB::empty();
+        let mut child_r_aabb = AABB::empty();
+        for i in 0..(NUM_BUCKETS - 1) {
```

Comment:
You could, BUT it would bring no merit at all, because you would have to recompute the split in line 187 anyway. You would have to fold on `(0..NUM_BUCKETS-1)` which can be expressed better with a loop.